### PR TITLE
feat: add linkTags API

### PR DESCRIPTION
**Pull request by johnnyreilly**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [x] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

I'm trying to add a custom link tag to the header of all the pages on my Docusaurus site. In my case I'm looking to render a web monetization link which looks like this:

```html
<link rel="monetization" content="https://ilp.uphold.com/LwQQhXdpwxeJ" />
```

It looks like there isn't a way to generically supply global link tags with Docusaurus. At least... until now!

I've implemented a new API called `linkTags` which is a generic mechanism for rendering `link` tags. It is modelled after the `metadata` API and sits alongside it in the `ThemeConfig` of `docusaurus.config.js`

## Test Plan

I'm not sure where unit tests for this would go - but if you give me a pointer I'll write one!  In the meantime you can see a screenshot of this working in this PR and see it working in the preview website also.

### Test links

image

Look at that!  It works! 

Deploy preview: https://deploy-preview-8077--docusaurus-2.netlify.app/

Crack open the devtools and you can see the `link` tag present and correct.

## Related issues/PRs

https://github.com/facebook/docusaurus/issues/8049

### CODE REVIEW

**Comment by Josh-Cena** [comment]
I don't have any opinions on this, but I wonder whether we should make it a theme API (like `metadata`) or a core API (like `stylesheets`). I don't really know why we even have the latter.

**Comment by johnnyreilly** [comment]
I put it where it is as it seems like the logical bedfellow to `metadata`. It would feel perhaps a little strange if you globally configured `meta` tags in one place and `link` tags in another. But I don't feel super strongly

**Comment by Josh-Cena** [comment]
Me neither. Feels weird since `stylesheets` does quite the same thing, but I can see why it lives in core (we might do some other optimizations with them).

File: website/docusaurus.config.js [hunk 1]
```diff
@@ -408,6 +408,8 @@ const config = {
       },
       image: 'img/docusaurus-soc.png',
       // metadata: [{name: 'twitter:card', content: 'summary'}],
+      linkTags: [{rel:"monetization", content:"https://ilp.uphold.com/LwQQhXdpwxeJ"}],
+      // if used would create <link rel="monetization" content="https://ilp.uphold.com/LwQQhXdpwxeJ" >
```
Comment (Josh-Cena) [inline website/docusaurus.config.js:412]:
```suggestion
      linkTags: [
        {
          rel: 'monetization',
          content: 'https://ilp.uphold.com/LwQQhXdpwxeJ',
        },
      ],
```

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
Looks almost good! Could you add some docs in https://docusaurus.io/docs/api/themes/configuration and possibly https://docusaurus.io/docs/seo?

**Comment by johnnyreilly** [comment]
Awesome - will do!  Just heading to bed now; will put some docs up tomorrow I expect.

**Comment by johnnyreilly** [comment]
> Could you add some docs in https://docusaurus.io/docs/api/themes/configuration and possibly https://docusaurus.io/docs/seo?

Added docs:

- https://deploy-preview-8077--docusaurus-2.netlify.app/docs/api/themes/configuration#linktags
- https://deploy-preview-8077--docusaurus-2.netlify.app/docs/seo

**Comment by johnnyreilly** [comment]
Paging @Josh-Cena 😅

File: website/docs/seo.md [hunk 2]
```diff
@@ -14,20 +14,24 @@ Docusaurus supports search engine optimization in a variety of ways.
...
-Provide global meta attributes for the entire site through the [site configuration](./configuration.md#site-metadata). The metadata will all be rendered in the HTML `<head>` using the key-value pairs as the prop name and value.
+Provide global meta and / or link tags for the entire site through the [site configuration](./configuration.md#site-metadata). The metadata and / or link tags will all be rendered in the HTML `<head>` using the key-value pairs as the prop name and value.
 
 ```js title="docusaurus.config.js"
 module.exports = {
   themeConfig: {
     metadata: [{name: 'keywords', content: 'cooking, blog'}],
     // This would become <meta name="keywords" content="cooking, blog"> in the generated HTML
+    linkTags: [
+      {rel: 'monetization', content: 'https://ilp.uphold.com/LwQQhXdpwxeJ'},
```
Comment (sublimator) [inline website/docs/seo.md:25]: `<link rel="monetization" >` tags encode the payment endpoint inside  the href attribute

Comment (johnnyreilly) [reply website/docs/seo.md]: Good catch! Fixed

**Comment by johnnyreilly** [comment]
There's a failing Windows test, but by the looks of it, it's a flaky test and not linked to this PR. My guess is that if you rerun it'll pass.

**Comment by slorber** [comment]
What prevents you from using a plugin instead? 

The API is not easy to discover but I think this impl should be good enough?

```js
function pluginMonetization(context, options) {
  return {
    name: "plugin-monetization",
    injectHtmlTags() {
      return {
        headTags: [
          {
            tagName: "link",
            attributes: {
              rel: "monetization",
              content: options.monetizationContent
            }
          }
        ]
      };
    }
  };
}
```

---

Although I think we discussed in the past it might be convenient to add a such shortcut to add global site links more conveniently, I'm not sure it's the responsibility of the theme to do so, as it would be duplicated in another theme that we'd want to implement.

Maybe it's better to add this to `createBootstrapPlugin()` that is always added to any Docusaurus site in core (used to provide similar shortcuts?

```js
export function createBootstrapPlugin({
  siteDir,
  siteConfig,
}: LoadContext): LoadedPlugin {
  const {
    stylesheets,
    scripts,
    clientModules: siteConfigClientModules,
  } = siteConfig;
  return {
    name: 'docusaurus-bootstrap-plugin',
    content: null,
    options: {
      id: 'default',
    },
    version: {type: 'synthetic'},
    path: siteDir,
    getClientModules() {
      return siteConfigClientModules;
    },
    injectHtmlTags: () => {
      const stylesheetsTags = stylesheets.map(
        (source): string | HtmlTagObject =>
          typeof source === 'string'
            ? `<link rel="stylesheet" href="${source}">`
            : {
                tagName: 'link',
                attributes: {
                  rel: 'stylesheet',
                  ...source,
                },
              },
      );
      const scriptsTags = scripts.map((source): string | HtmlTagObject =>
        typeof source === 'string'
          ? `<script src="${source}"></script>`
          : {
              tagName: 'script',
              attributes: {
                ...source,
              },
            },
      );
      return {
        headTags: [...stylesheetsTags, ...scriptsTags],
      };
    },
  };
}
```

---

Also not a fan of "linkTags" naming. We already have stylesheets, scripts and metadata (which afaik is plural of metadatum) so "links" makes more sense to me (also similar to Remix: https://remix.run/docs/en/v1/api/conventions#links)

--- 

Side note but stylesheets + scripts are in core/siteConfig, and metadata is in the theme/themeConfig, that's a bit weird

https://docusaurus.io/docs/api/themes/configuration#metadata

Maybe we'll want to add a site.metadata in core too for consistency? 🤷‍♂️ 

Some important thing to consider: we set some global non-core opinionated meta in the theme currently (I think just `<meta name="twitter:card" content="summary_large_image" />`) and users should be able to easily override this on a global site level. Declaration order of metadata will matter to keep this behavior.

**Comment by slorber** [comment]
What about just using the synthetic boottrap plugin above to provide a `siteConfig.head` or `siteConfig.headTags` API?

Like suggested here: https://github.com/facebook/docusaurus/issues/8049#issuecomment-1239493341

In the end it's the most flexible way to add any tag in the head, no matter what it is. anything else would just be a shortcut.

I'd rather start with more generic, less convenient apis first, and then add shortcuts like `linkTags` later (eventually). That the philosophy of the extensible web manifesto (https://extensiblewebmanifesto.org/).

Note that we already have a low-level api (plugin), but agree this shortcut (`siteConfig.head`) would be convenient.

**Comment by johnnyreilly** [comment]
> The API is not easy to discover but I think this impl should be good enough?

It works - but this is hard to discover and very verbose for what you're trying to achieve (add some link tags).  Feels like you have to buy a car when you only want to drive a mile 😄 

> Also not a fan of "linkTags" naming. We already have stylesheets, scripts and metadata (which afaik is plural of metadatum) so "links" makes more sense to me (also similar to Remix: https://remix.run/docs/en/v1/api/conventions#links)

Happy with `links` as a name; would be happy to switch.

> What about just using the synthetic boottrap plugin above to provide a `siteConfig.head` or `siteConfig.headTags` API?
> 
> Like suggested here: https://github.com/facebook/docusaurus/issues/8049#issuecomment-1239493341

I'd be happy to have a go at this - do you have any examples in the codebase of how this is done already?  I can't see any piping examples in the codebase so far?

**Comment by slorber** [comment]
Similarly, it can also be handled in the bootstrap plugin:


```js
export function createBootstrapPlugin({
  siteDir,
  siteConfig,
}: LoadContext): LoadedPlugin {
  const {
    stylesheets,
    scripts,
    head, // <==== NEW
    clientModules: siteConfigClientModules,
  } = siteConfig;
  return {
    name: 'docusaurus-bootstrap-plugin',
    // other things
    injectHtmlTags: () => {
      const stylesheetsTags = [];
      const scriptsTags = [];
      return {
        headTags: [
          ...head, // <==== NEW
          ...stylesheetsTags, ...scriptsTags],
      };
    },
  };
}
```


Does it make sense?

**Comment by johnnyreilly** [comment]
Sorry no - I don't follow!  I don't know what the `createBootstrapPlugin` does.... And I'm not sure what consumption would then look like in a `docusaurus.config.js`?

**Comment by slorber** [comment]
> I'm guessing that all users are somehow using the createBootstrapPlugin and they just don't know it?

Exactly: all sites implicitly have this plugin by default.

Use the code above where I added the NEW comments, and users will be able to use in siteconfig:

`{head: {tag: "link", attributes: {rel: "monetization"}}}`

**Comment by johnnyreilly** [comment]
okay cool and they'd do it at the same level as `stylesheets` by the looks of it?

I think I could have a crack at this.  Probably at the weekend.

**Comment by johnnyreilly** [comment]
Cool - I think I get it. I'll have a go and report back!

**Comment by johnnyreilly** [comment]
Closed in favour of @slorber suggested alternate implementation: https://github.com/facebook/docusaurus/pull/8151

**Pull request closed without merging**